### ISSUE TITLE
[BUG] dex.trades missing >50 equalizer LPs

### ISSUE THREAD

**Issue by ftm1337** [duneanalytics/spellbook#3200]
### Description

Missing Data in `dex.trades`: Around 50 Equalizer.exchange pools are missing.

### Current behavior

Only around 146 pairs out of ~200 pairs are present in the dex.trades table for equalizer, including some of the biggest ones.  

### Expected behavior

- If all the pools are included, Equalizer's volume should account for 30-40% of all volumes on Fantom.
- This query shows an example of a missing pool: https://dune.com/queries/2386231/3912771?sidebar=none
- Here is what it should look like: https://defillama.com/dexs/equalizer-exchange

*Might be unrelated but `dex.trades` doesnt have names on prominent tokens like TAROT, EQUAL, MPX, MUMMY, etc. and other bigger ones like BIFI, DEUS, fBOMB dont have `amount_usd` values https://dune.com/queries/2385999/3912732*

### Impacted model(s)

```
dex.trades
```

### Possible solution

- Possible misconfiguration: Same 2 tokens can have more than 1 liquidity pool?
- This query returns 148 rows: 
  - SELECT distinct contract_address FROM equalizer_exchange_fantom.**Pair_evt_Swap**
- BUT, This query returns 198 rows: 
  - SELECT distinct contract_address FROM equalizer_exchange_fantom.**PairFactory_evt_PairCreated**

**Comment by jeff-dude** [comment]
@ftm1337 thanks for raising. starting simple: have you seen the code to build affected data [here](https://github.com/duneanalytics/spellbook/blob/main/models/equalizer_exchange/fantom/equalizer_exchange_fantom_trades.sql)? you can leverage that as a starting point to narrow down potentially why some pairs are missing

**Comment by jeff-dude** [comment]
as for token metadata, such as symbol/decimals, please use [this file](https://github.com/duneanalytics/spellbook/blob/main/models/tokens/fantom/tokens_fantom_erc20.sql). once added, they will propagate through to `dex.trades`

**Comment by ftm1337** [comment]
@jeff-dude is it possible that the issue is arising from the start date being set to 23-nov-2023 instead of the **actual** 03-nov-2023 ?
i think pools that were created between 3-23 were skipped..

**Comment by jeff-dude** [comment]
yep, for sure. if the start date is late, then all preceding data will be missing. feel free to submit a PR with correct start date. the PR will run a gh action which generates a test table of the affected models in PR, which will be live on dune for ~24 hours for you to query and test

**Comment by ftm1337** [comment]
@jeff-dude the missing pools still dont show up..
```
select distinct project_contract_address 
from dex.trades 
where project = 'equalizer_exchange'
```
does not contain all pools.. it has only about 60% of pools..

**Comment by jeff-dude** [comment]
the first step i'd recommend is to take the latest code from `main` branch, run `dbt compile` to build sql code, find the model under the target directory and manually debug in the dune app. you can comment out code to start, keep it simple with the decoded table and filter on missing pools, ensure it returns as expected, then uncomment code in the spell until the result disappears. see if that helps find the bug

**Comment by jeff-dude** [comment]
@ftm1337 
if you look at the spell that loads equalizer data into `dex.trades`
https://github.com/duneanalytics/spellbook/blob/main/models/equalizer_exchange/fantom/equalizer_exchange_fantom_trades.sql

we can see the same decoded tables you're querying.
however, it looks like the pools you're reporting as missing don't come up in swaps decoded table.
here is example query that shows the pairs created, but when joined to swap, it drops out.
```
with
  dex as (
    SELECT
      *
    FROM
      dex."trades" AS t
    WHERE
      project = 'equalizer_exchange'
      AND project_contract_address = 0x7547d05dff1da6b4a2ebb3f0833afe3c62abd9a1
  ),
  pair_created as (
    select
      *
    FROM
      equalizer_exchange_fantom.PairFactory_evt_PairCreated f
    where
      f.pair = 0x7547d05dff1da6b4a2ebb3f0833afe3c62abd9a1
  ),
  spell_logic as (
    select
      f.*
    FROM
      equalizer_exchange_fantom.Pair_evt_Swap t
      inner join equalizer_exchange_fantom.PairFactory_evt_PairCreated f on f.pair = t.contract_address
    where
      t.contract_address = 0x7547d05dff1da6b4a2ebb3f0833afe3c62abd9a1 --this field populates the dex column
  )
SELECT
  'evt_PairCreated',
  *
from
  pair_created
union all
SELECT
  'evt_Swap',
  *
from
  spell_logic
LIMIT
  1
```

is it possible they have no swaps?
if not, are you missing contracts to be decoded and added to the spell linked above?

from what i see, there is no issue with the code that loads `dex.trades`. the issue, if any, would be on missing contracts in decoded layer.

**Comment by ftm1337** [comment]
Hey @jeff-dude! thanks for fixing this :D
Please send me your public evm address (i am `@i543` on Discord if you prefer that), as we had appointed a $100 bounty while opening this issue.
Thank you once again <3

**Comment by jeff-dude** [comment]
> Hey @jeff-dude! thanks for fixing this :D Please send me your public evm address (i am `@i543` on Discord if you prefer that), as we had appointed a $100 bounty while opening this issue. Thank you once again <3

interesting!
my public evm is `jdude.eth`

**Comment by ftm1337** [comment]
hey jeff, looks like the issue still persists :sweat_smile:  https://dune.com/queries/3070210

**Comment by jeff-dude** [comment]
> hey jeff, looks like the issue still persists 😅 https://dune.com/queries/3070210

looks like `equalizer` project has been updated, naming standard wise:
```
WITH 
t1 AS (SELECT DISTINCT contract_address FROM equalizer_exchange_fantom.Pair_evt_Swap)
, t2 AS (SELECT DISTINCT project_contract_address FROM dex.trades WHERE project = 'equalizer')
, t3 AS (SELECT DISTINCT pair FROM equalizer_exchange_fantom.PairFactory_evt_PairCreated)
, t4 AS (SELECT DISTINCT contract_address FROM equalizer_exchange_fantom.Pair_evt_Initialized)
--, t5 AS (SELECT _0 AS id,pair AS pool_id,stable,(token0||token1) AS pair_id,token0,token1,1 AS one FROM equalizer_exchange_fantom.PairFactory_evt_PairCreated)
--, t6 AS (SELECT t5.id, t4.contract_address AS dune_init_pools,t5.pool_id AS all_pools FROM t5 LEFT OUTER JOIN t4 ON t5.pool_id = t4.contract_address)


SELECT 'Pair_evt_Swap' as tableName, COUNT(*) as numRows FROM t1
UNION ALL SELECT 'dex.trades_project=equalizer_exchange' as tableName, COUNT(*) as numRows FROM t2
UNION ALL SELECT 'PairFactory_evt_PairCreated' as tableName, COUNT(*) as numRows FROM t3
UNION ALL SELECT 'Pair_evt_Initialized' as tableName, COUNT(*) as numRows FROM t4
```

**Comment by ftm1337** [comment]
i see.. yes, its now `equalizer` :+1: 
but still, the results are similar.. `PairFactory_evt_PairCreated` is logged **367 times** by dune (emitted by the factory) and the `Pair_evt_Initialized` is logged just **157 times** (emitted by the created pool right at the time of creation)

i think i found the solution.. im going to submit a few pool contracts to dune for parsing, im guessing there's a difference in bytecode that makes some pairs unavailable..

**Comment by jeff-dude** [comment]
> i see.. yes, its now `equalizer` 👍 but still, the results are similar.. `PairFactory_evt_PairCreated` is logged **367 times** by dune (emitted by the factory) and the `Pair_evt_Initialized` is logged just **157 times** (emitted by the created pool right at the time of creation)
> 
> i think i found the solution.. im going to submit a few pool contracts to dune for parsing, im guessing there's a difference in bytecode that makes some pairs unavailable..

yep, that's what i was working towards in previous comments. looks like missing decoded contracts to add to the spell. let me know if you get those added and i can help merge 👍

**Comment by ftm1337** [comment]
i think there is more to it.. :thinking: 
(referring to this same query, https://dune.com/queries/3070210/5113730?sidebar=none)

Consider this : As expected, if a contract isnt decoded, it's pain_intialized event isnt picked up
![image](https://github.com/duneanalytics/spellbook/assets/84405345/e6bd2788-db1c-42c7-9193-c832fb0f301c)

But interestingly, some contracts have been decoded by dune but their corresponding Pair_Initialized event isnt picked up..
![image](https://github.com/duneanalytics/spellbook/assets/84405345/a5cc24a4-d252-49f3-889f-56e2f89a6b5a)

**Comment by discochuck** [comment]
@jeff-dude 

Here's my work trying to debug this:

https://dune.com/queries/4200994

The query it self isn't doing much, but this is from the commented out bits as I was trying to sus out the issue. 
--snip

My theory was that the old Pair contract in the old 'equalizer' namespace would account for most if not all
of the missing volume, and thus that would need to be unioned (or whichever would make sense) with the 
https://github.com/duneanalytics/spellbook/blob/main/dbt_subprojects/dex/models/_projects/equalizer/fantom/equalizer_fantom_pools.sql
of it all

This was one of the big pools/pairs that had 500k volume in the last 24h that showed up in the old namespace but wouldn't be caught
by the Swap event in the new equalizer_exchange namespace. 

0x3d6c56f6855b7cc746fb80848755b0a9c3770122 - Equal      500k last 24h
0x670058c09c51f91a1b015aee967858fb07aa98cf - Goglz      2.0M last 24h
0x997fbfb238125db4e90a62b423bbed2e202edd5a - Tango      147k last 24h

I also assumed that any Swap event from the old and busted namespace would not match a tx_hash from the newhottness 

This was wrong, but it seems the equalizer_fantom.Pair_evt_Swap events are capturing some of the larger volume pairs 

I don't understand why swaps from the Pair in the old namespace would exist at all in the swaps in the new namspace. 

---

**Comment by discochuck** [comment]
Guess I forgot to put the ask in there: 

Could we get the equalizer_fantom.Pair_evt_Swap included in dex trades spell for equalizer_echange ?

**Comment by jeff-dude** [comment]
> Guess I forgot to put the ask in there:
> 
> Could we get the equalizer_fantom.Pair_evt_Swap included in dex trades spell for equalizer_echange ?

for sure. i personally don't have time currently to make the change, but you can modify this spell to use new equalizer schema name:
https://github.com/duneanalytics/spellbook/blob/main/dbt_subprojects/dex/models/trades/fantom/platforms/equalizer_fantom_base_trades.sql

or alternatively, you could add the macro call twice and union the rows together if necessary. to do this, wrap both in CTE each, then union outside of them.

**Comment by jeff-dude** [comment]
closing issue due to inactivity. if still relevant, and proposed fix is available, feel free to reopen an issue as needed.
